src/apps/replication/lib/mutation_log.cpp

```diff
@@ -564,7 +565,8 @@ void mutation_log::internal_write_callback(
 
     if (err == ERR_OK || err == ERR_HANDLE_EOF)
     {
-        dassert(g_end_offset == end_offset,
+        // the log may still be written when used for learning
```

Comment:
when a secondary replica does checkpoint, it will stall itself from committing the local app. When checkpoint is done, it has to catchup using local private log when the prepare list is not enough to hold all these committed decrees during checkpointing. At the same time when catching up, writing still goes on which appends the private log.